### DISCUSSION TITLE
Trial-level metadata add and save externally

### DISCUSSION THREAD

**Discussion by anons9** [optuna/optuna#3812] [Q&A] [answered]
Hello,

I am using optuna in combination with neptune.ai, however, in the study-level run, optuna only saves its own parameters about the trials (in folders 0,1,2,etc.). There, I would also like to save some of my own trial-level metadata (for example, a plot I made about that trial). What would be the best way to approach this?

One way is suggested by Neptune.ai CMO Jakub in this video: https://docs.neptune.ai/integrations-and-supported-tools/hyperparameter-optimization/optuna, where he creates a separate run for each trial. However, that can get really messy with, e.g., a 1000 trials (even with grouping by sweep id). 

Should I open an issue on Optuna asking for this feature?

Thank you in advance!

**Comment by nzw0301** [comment, marked as answer]
I'm not familiar with neptune.ai.  But to save a trial-level value, we can use user attributes as in https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/003_attributes.html#adding-user-attributes-to-trials. For the plot case, I might save figure on local storage or as byte object and save the file path or byte object as a trial attribute.
